### PR TITLE
feat(images): replace kubectl and helm-release-upgrader with distroless kubescape-ops-tool

### PR DESCRIPTION
**Pull request by matthyx**
## Overview

Supersedes #928 with a different approach. Adds a single, fully non-debug-distroless image — `quay.io/kubescape/kubescape-ops-tool` — built `golang:1.26-trixie` (builder) → `gcr.io/distroless/static-debian13:nonroot` (final), mirroring [kubevuln's `build/Dockerfile`](https://github.com/kubescape/kubevuln/blob/main/build/Dockerfile). It replaces **both** `quay.io/kubescape/kubectl` and `quay.io/kubescape/helm-release-upgrader` across the whole `kubescape-operator` chart.

Where #928 kept an Alpine-based image with bash/kubectl/jq/openssl (because the chart's usages shell out), this PR instead reimplements every one of those usages natively in Go, so the final image needs no shell and no CLI binaries at all:

| Subcommand | Replaces | Used by |
|---|---|---|
| `certgen-create` | `scripts/certgen-create.sh` | storage + operator admission-webhook init containers |
| `certgen-patch` | `scripts/certgen-patch.sh` | same |
| `rollout-restart` | `kubectl rollout restart` via `sh -c` | grype-offline-db cronjob |
| `helm-upgrade` | `helm-release-upgrader`'s bash entrypoint (via `dtzar/helm-kubectl`) | autoupdater cronjob |

Full design rationale, alternatives considered, and risk analysis: `docs/features/kubescape-ops-tool.md` in this PR, and the underlying spec/plan (produced via a deep-interview + Planner/Architect/Critic consensus pass) at `.omc/specs/deep-interview-kubescape-ops-tool.md` / `.omc/plans/kubescape-ops-tool.md` (not committed — local planning artifacts).

## Verification performed

- `go build ./...`, `go vet ./...`, `go test ./... -race` — all green, includes an exact-byte assertion on the certgen CA-bundle patch, a test that `helm-upgrade` has no hardcoded defaults, and a regression test proving `action.Upgrade.ResetThenReuseValues` (not `ReuseValues`) is required so an upgrade from a pre-1.41.0 release still picks up the new `kubescapeOpsTool` values tree.
- `docker buildx build` succeeds; `docker run --help` lists the 4 subcommands; `--entrypoint /bin/sh ... -c true` **fails** (no shell); works under `--read-only --user 65532:65532` and `--user 100:100`.
- `trivy image` on the built tar: **0 CRITICAL** findings, no OS packages (distroless).
- Image tar contains no `sh`/`bash`/`busybox`/`kubectl`/`helm`/`jq`/`openssl`.
- `helm lint` clean; `helm unittest charts/kubescape-operator`: 16 suites / 119 tests / 1060 snapshots, all pass.

## Breaking change

`values.yaml`'s `certificates.certgen.image`, `grypeOfflineDB.rollout.image`, and `helmReleaseUpgrader.image` are removed (no deprecation shim) in favor of one shared `kubescapeOpsTool.image` block. `Chart.yaml` bumped to `1.41.0`. Details in `docs/features/kubescape-ops-tool.md`.

## Sequencing note for reviewers

The image (`images/kubescape-ops-tool/`) and CI workflow in this PR need to be built, scanned, and published to `quay.io/kubescape/kubescape-ops-tool:0.1.0-r1` (via the `kubescape-ops-tool-image.yaml` workflow's manual `publish` gate, restricted to `refs/heads/main`) **before merge**, so the chart's pinned tag resolves. Auth uses the org's `QUAYIO_REGISTRY_USERNAME`/`QUAYIO_REGISTRY_PASSWORD` secrets directly (same pattern as `kubescape/kubevuln`'s publish workflows) — no separate GitHub Environment is required. A disposable per-commit `test-<sha>` image is already pushed automatically on every push to this PR (restricted to same-repo branches) for pre-merge testing.

## Related issues/PRs

* Supersedes #928
* Relates to #923

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added the `kubescape-ops-tool` image for TLS certificate management, deployment restarts, and Helm upgrades.
  - Updated the operator chart to use the shared operations image across certificate, database, and autoupdater tasks.
  - Added configurable Helm upgrade settings and improved non-root execution for the autoupdater.

- **Improvements**
  - Updated the chart and application version to 1.41.0.
  - Added automated image build, security scanning, testing, and publishing workflows.
  - Added Helm tests covering image usage, command behavior, configuration overrides, and security settings.
  - Added documentation for the operations image and autoupdater behavior.

### CODE REVIEW

**Comment by matthyx** [comment]
A ready-to-try build is up: \`quay.io/kubescape/kubescape-ops-tool:test-f2d1202\` (multi-arch, amd64+arm64). Every push to this branch pushes a fresh \`test-<sha>\` tag automatically, so no need to wait for merge to try it out, e.g.:

\`\`\`
docker run --rm quay.io/kubescape/kubescape-ops-tool:test-f2d1202 --help
\`\`\`

File: images/kubescape-ops-tool/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+WORKDIR /work
+ARG TARGETOS TARGETARCH
+
+COPY go.mod go.sum ./
+RUN go mod download
+
+COPY . .
+RUN GOOS=$TARGETOS GOARCH=$TARGETARCH go build -trimpath -ldflags="-s -w" -o /out/kubescape-ops-tool .
+
+FROM gcr.io/distroless/static-debian13:nonroot@sha256:e2e927ec666bae08560abb3c55d0659eceabb657f56b6782ab500a9fc7f555e3
```
Comment (matthyx) [inline images/kubescape-ops-tool/Dockerfile]: I'd rather use a dynamic tag:

```suggestion
FROM gcr.io/distroless/static-debian13:nonroot
```

File: .github/workflows/kubescape-ops-tool-image.yaml [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+          path: kubescape-ops-tool-${{ matrix.arch }}.tar
+          retention-days: 2
+
+  pr-test-image:
+    needs: build-test-scan
+    if: github.event_name == 'pull_request'
+    runs-on: ubuntu-latest
+    env:
+      QUAY_USERNAME: ${{ secrets.QUAYIO_REGISTRY_USERNAME }}
+      QUAY_PASSWORD: ${{ secrets.QUAYIO_REGISTRY_PASSWORD }}
```
[the remark this answers is not in the document]
Comment (matthyx) [reply .github/workflows/kubescape-ops-tool-image.yaml:94]: Fixed in bd297f0: `pr-test-image` now requires `github.event.pull_request.head.repo.full_name == github.repository`. GitHub already withholds secrets from fork-originated `pull_request` runs by default (the actual "PR author controls the workflow YAML" attack surface), but this makes that boundary explicit rather than relying on the platform default alone. Same-repo branch PRs — the only case this leaves eligible — already require push access to this repo to open in the first place, i.e. the same trust level as being able to edit any workflow file directly. Worth noting `kubescape/kubevuln`'s own `pr-image.yaml` uses this identical secrets-in-a-pull_request-job pattern without even this guard, so this hardens beyond the org's existing precedent.

File: images/kubescape-ops-tool/main.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+func runHelmUpgrade(args []string) error {
+	o, err := helmupgrade.ParseFlags(args, os.Getenv, os.Stderr)
+	if err != nil {
+		os.Exit(2)
+	}
+	settings := cli.New()
+	settings.SetNamespace(o.Namespace)
+
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(settings.RESTClientGetter(), o.Namespace, os.Getenv("HELM_DRIVER"), func(format string, v ...any) {
```
[the remark this answers is not in the document]
Comment (matthyx) [reply images/kubescape-ops-tool/main.go:123]: I don't think this is actually a bug — verified against the exact vendored source for the versions in `go.sum` (`k8s.io/client-go@v0.37.0`, `k8s.io/cli-runtime@v0.37.0`, `helm.sh/helm/v3@v3.22.0`):

`settings.RESTClientGetter()` returns a `*genericclioptions.ConfigFlags`. Its `ToRESTConfig()` → `toRawKubeConfigLoader()` builds a `clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, overrides)` (`cli-runtime@v0.37.0/pkg/genericclioptions/config_flags.go:255`). `cli.New()` sets `ConfigFlags.KubeConfig` to a pointer to an empty string when `$KUBECONFIG`/`--kubeconfig` aren't set, which leaves `loadingRules.ExplicitPath` empty and falls through to the loader's normal precedence.

`DeferredLoadingClientConfig.ClientConfig()` (`client-go@v0.37.0/tools/clientcmd/merged_client_builder.go:96-127`) explicitly falls back to `inClusterClientConfig` whenever the merged config is empty/default and `icc.Possible()` is true. `Possible()` (`client_config.go:668`) just checks for `KUBERNETES_SERVICE_HOST`/`PORT` env vars plus the mounted SA token file — exactly what the autoupdater CronJob pod has via its ServiceAccount. This is the identical mechanism `kubectl` itself relies on for in-cluster auth.

So `actionConfig.Init(settings.RESTClientGetter(), ...)` does pick up the pod's ServiceAccount automatically without any explicit `KUBECONFIG`/in-cluster wiring — no code change needed here. Happy to add an integration-style test against an in-cluster env if that would help close this out, but I didn't want to add code that duplicates behavior `cli-runtime` already provides.

File: .github/workflows/kubescape-ops-tool-image.yaml [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            docker load -i "kubescape-ops-tool-${arch}.tar"
+            docker tag "kubescape-ops-tool-test:${arch}" "${IMAGE}:${TEST_TAG}-${arch}"
+            docker push "${IMAGE}:${TEST_TAG}-${arch}"
+          done
+          docker buildx imagetools create -t "${IMAGE}:${TEST_TAG}" "${IMAGE}:${TEST_TAG}-amd64" "${IMAGE}:${TEST_TAG}-arm64"
+          echo "Pushed disposable test image: ${IMAGE}:${TEST_TAG}" | tee -a "$GITHUB_STEP_SUMMARY"
+
+  publish:
+    needs: build-test-scan
+    if: github.event_name == 'workflow_dispatch' && inputs.publish && github.ref == 'refs/heads/main'
```
[the remark this answers is not in the document]
Comment (matthyx) [reply .github/workflows/kubescape-ops-tool-image.yaml:123]: This is intentional, not a bug — the `main`-only gate exists specifically so the chart's pinned `kubescapeOpsTool.image.tag` can never reference a tag that doesn't exist yet. The actual sequencing (documented in the PR description) is: merge this PR (image + workflow + chart together) → run `workflow_dispatch` with `publish=true` on `main` → the `0.1.0-r1` tag now exists → the chart, already merged, resolves it correctly.

For pre-merge testing without touching that guarded release tag, every push to this PR already auto-publishes a disposable `test-<sha>` multi-arch image via the `pr-test-image` job (now also same-repo-restricted, see the sibling thread) — that's the intended pre-merge path, not the guarded `0.1.0-r1` release tag.

File: .github/workflows/kubescape-ops-tool-image.yaml [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (matthyx) [reply .github/workflows/kubescape-ops-tool-image.yaml:124]: This finding is against the PR description's older sequencing note, which I've now updated — no `quay-kubescape-ops-tool` GitHub Environment is used or needed. The publish job authenticates via `docker/login-action` using the org-level `secrets.QUAYIO_REGISTRY_USERNAME`/`QUAYIO_REGISTRY_PASSWORD` directly (same pattern as `kubescape/kubevuln`'s own publish workflows), not an environment-scoped secret. Those are already-provisioned org secrets, confirmed working against the real `quay.io/kubescape/kubescape-ops-tool` repo (the `pr-test-image` job has already pushed several real tags there).

File: images/kubescape-ops-tool/Dockerfile (hunk 1 above)
Comment (matthyx) [reply images/kubescape-ops-tool/Dockerfile]: Done in bd297f0. Also removed `kubescape-ops-tool-base-refresh.yaml` since its only purpose was bumping the now-removed pinned digest.

File: .github/workflows/kubescape-ops-tool-image.yaml [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    needs: build-test-scan
+    # Registry secrets must never be reachable from a workflow whose YAML a PR
+    # author controls. GitHub already withholds secrets from fork-originated
+    # `pull_request` runs by default; this condition makes that boundary
+    # explicit rather than relying on the platform default alone. Same-repo
+    # branch PRs (the only case this leaves eligible) already require push
+    # access to this repository to open in the first place.
+    if: >-
+      github.event_name == 'pull_request' &&
+      github.event.pull_request.head.repo.full_name == github.repository
```
[the remark this answers is not in the document]
Comment (matthyx) [reply .github/workflows/kubescape-ops-tool-image.yaml:90]: Correct catch — my same-repo guard only excludes fork PRs, which never had secrets to begin with on a plain `pull_request` trigger. It does nothing for the actual risk: a same-repo branch PR can modify this workflow's YAML and have it run with real Quay credentials on the next push, no review required first.

Leaving this as-is deliberately rather than fixing further: `kubescape/kubevuln`'s `pr-image.yaml` has the identical pattern (`QUAYIO_REGISTRY_USERNAME`/`PASSWORD` at job level, `pull_request` trigger, no environment gate, no protected-ref requirement) already live in production. This PR isn't introducing a new or worse risk than what's already accepted org-wide for exactly this use case (letting reviewers pull a disposable per-commit test image before merge) — it's matching existing practice. If the org wants to tighten this, it'd be a change to both workflows, not something to solve differently just in this one.

File: .github/workflows/kubescape-ops-tool-image.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (matthyx) [reply .github/workflows/kubescape-ops-tool-image.yaml:94]: Correction to my own reply above: CodeRabbit's follow-up review (below) correctly points out this guard doesn't actually narrow the trust boundary — it only excludes fork PRs, which never had secrets on `pull_request` anyway. Same-repo branch PRs (the case that matters) are still fully exposed. See my reply on that thread for why I'm leaving it as-is: it matches `kubescape/kubevuln`'s existing `pr-image.yaml` pattern exactly, not a new risk.

File: images/kubescape-ops-tool/internal/helmupgrade/upgrade.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+	repoFile.Update(entry)
+	if err := repoFile.WriteFile(settings.RepositoryConfig, 0o644); err != nil {
+		return nil, fmt.Errorf("write %s: %w", settings.RepositoryConfig, err)
+	}
+	return index, nil
+}
+
+// Upgrade resolves the chart from the registered repo and upgrades the release.
+func Upgrade(cfg *action.Configuration, settings *cli.EnvSettings, o Options) (*release.Release, error) {
+	if _, err := AddRepo(settings, o.ChartRepo, o.ChartRepoURL); err != nil {
```
Comment (amirmalka) [inline images/kubescape-ops-tool/internal/helmupgrade/upgrade.go]: **The release-exists guard from the script being replaced is gone, and the script documents exactly why it was there.**

`kubescape/helm-release-upgrader`'s `entrypoint.sh` opens with:

```bash
if helm status kubescape ; then
	echo "Kubescape is installed, starting update..."
else
	echo "Kubescape Helm release not found, exiting silently."
	exit 0
fi
```

and its header comment spells out the scenario: *"The CronJob does NOT get cleaned up in the cluster and might run even when there is no Kubescape release"* — hence *"exit silently to not fail the CronJob and avoid pointless retries."*

`Upgrade` has no equivalent. It calls `AddRepo`, locates the chart, and goes straight to `up.Run(o.Release, ch, …)`, which returns `release: not found` when the release is absent. That propagates up through `runHelmUpgrade` to `main.go`'s `slog.Error` + `os.Exit(1)`, the container exits non-zero, and `restartPolicy: OnFailure` retries it.

**Failure scenario.** A user uninstalls the kubescape release but the CronJob object survives — the precise case the old comment calls out. Before: the job runs, logs one line, exits 0, and nobody notices. After: every scheduled run fails, retries under `OnFailure`, and the CronJob sits permanently failed with `release: not found`. Scoped to `capabilities.autoUpgrading: enable` clusters, which are by definition the ones nobody is watching.

A `histories, err := action.NewHistory(cfg).Run(o.Release)` (or `NewGet`) check at the top of `Upgrade`, returning nil on `driver.ErrReleaseNotFound`, reproduces the old contract in about five lines.

File: images/kubescape-ops-tool/internal/helmupgrade/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	up.Namespace = o.Namespace
+	up.Version = o.Version
+	up.Timeout = o.Timeout
+	// An unattended upgrade must keep the user's prior values AND pick up any
+	// new keys the target chart's defaults introduce (e.g. a values tree that
+	// didn't exist in the installed release's chart version). ReuseValues alone
+	// would silently drop new chart defaults; ResetThenReuseValues loads the
+	// new chart's defaults first, then reapplies the user's previous values on
+	// top of them.
+	up.ResetThenReuseValues = true
```
Comment (amirmalka) [inline images/kubescape-ops-tool/internal/helmupgrade/upgrade.go]: **This is a genuine behavioural change to the autoupdater, and the comment justifies it against a baseline that was never in play.**

The comment reasons that *"ReuseValues alone would silently drop new chart defaults"*. That's true of `ReuseValues` — but `ReuseValues` isn't what the replaced tool did. `kubescape/helm-release-upgrader`'s entrypoint runs:

```
helm upgrade kubescape kubescape/kubescape-operator -n kubescape
```

No `--reuse-values`, no `--reset-then-reuse-values`, no `-f`, no `--set`. That's plain Helm semantics: the new chart's defaults are used and **every prior user override is discarded** on each run.

So the actual before/after is not "ReuseValues → ResetThenReuseValues", it's **"reset to chart defaults" → "chart defaults with prior user values layered back on top"**.

I think the new behaviour is the better one — silently wiping a user's customisation nightly is hard to defend. But it is a live change for exactly the clusters that never look: an override someone set once and assumed had been washed away will now survive, and keep surviving. That includes overrides targeting the three `image` keys this PR deletes, which will sit inert in the values map (harmless in Helm without a `values.schema.json`, but confusing to anyone reading `helm get values` afterwards).

Worth a line in `docs/features/kubescape-ops-tool.md`'s breaking-change section alongside the `securityContext` UID note, and worth rewording this comment so the next reader doesn't conclude the old tool preserved values when it did the opposite.

File: images/kubescape-ops-tool/internal/certgen/create.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+func generate(ctx context.Context, kc kubernetes.Interface, o CreateOptions) error {
+	slog.Info("secret not found, generating", "namespace", o.Namespace, "secret", o.SecretName)
+
+	hosts := splitHosts(o.Hosts)
+	if len(hosts) == 0 {
+		return fmt.Errorf("--host produced no DNS names")
+	}
+	notBefore := time.Now().Add(-time.Minute)
+	notAfter := notBefore.Add(time.Duration(o.Days) * 24 * time.Hour)
```
Comment (amirmalka) [inline images/kubescape-ops-tool/internal/certgen/create.go]: Minor: `--days` reaches this arithmetic with no validation.

`time.Duration` is `int64` nanoseconds, so it saturates at roughly 292 years — about 106,751 days. A `--days` above that overflows, `notAfter` lands *before* `notBefore`, and the tool cheerfully generates and persists an already-expired certificate, then reports success. `--days 0` or a negative value does the same thing more directly.

The 36500 default is comfortably safe, so this only bites someone who overrides it — but the failure is silent and the resulting secret is cached for every subsequent pod start via the reuse path, so the bad cert outlives the run that produced it. A `if o.Days <= 0 || o.Days > 36500*2 { return error }` in `ParseCreateFlags` closes it; the old shell had no guard either, but openssl would at least have refused rather than issuing a backwards validity window.

**Review by amirmalka — COMMENTED** [review]
Reviewed the whole change: the Go reimplementation against the two shell scripts it replaces, the chart wiring, and the publish workflow.

**The certgen rewrite is stricter than the shell it replaces, and I checked that rather than assuming it.** The CA template sets `IsCA`, `BasicConstraintsValid` and `KeyUsageCertSign|CRLSign` explicitly, where `openssl req -x509` was relying on whatever the ambient `openssl.cnf` default section provided. Serials are 128-bit random rather than `-CAcreateserial`. P-256 on both keys matches `prime256v1`; the leaf's CN, DNS SANs, `keyUsage` and `serverAuth` EKU are an exact match for the old `leaf.cnf`; and the private key is written `0600` instead of at the process umask. The `notBefore = now - 1m` skew tolerance is a real improvement over the shell too.

I checked the two things most likely to break a distroless conversion, and both are handled: `HELM_CACHE_HOME`/`HELM_CONFIG_HOME`/`HELM_DATA_HOME` are redirected into the emptyDir so Helm still works under `readOnlyRootFilesystem: true`, and the autoupdater's UID/GID/fsGroup moved 1000 → 65532 to match the distroless `nonroot` user. The chart swap itself is clean like-for-like — identical `securityContext` and args, with the script ConfigMaps and their `checksum/...-certgen-scripts` pod annotations removed consistently rather than left dangling.

The publish workflow's security model is right, which is worth saying explicitly since it handles registry credentials: it triggers on `pull_request` rather than `pull_request_target`, and the test-image push is gated on `github.event.pull_request.head.repo.full_name == github.repository`, so a fork PR can never reach `QUAYIO_REGISTRY_*`. Publish is gated on `workflow_dispatch` + explicit `publish` input + `refs/heads/main`.

One non-finding I chased and dropped: `certgen.Create` doesn't handle `AlreadyExists` on the secret-create path, but storage is `replicas: 1` and the two certgen init containers target different secrets, so there's no reachable race — and the shell behaved identically anyway.

Three findings inline. The first is a behavioural regression against the script being replaced, and I'd want it in before the `0.1.0-r1` tag is cut, since your sequencing note has the image published ahead of merge and re-publishing that tag afterwards is worse than getting it right the first time.

---

**One thing the review turned up that's worth recording even though it isn't a finding.**

I went in expecting an upgrade-path problem: chart 1.41.0 requires `.Values.kubescapeOpsTool`, which doesn't exist in 1.40.x, and the 1.40 → 1.41 hop is performed by the *old* upgrader, not this one — so the new `ResetThenReuseValues` can't help it. If the old tool used `--reuse-values`, that render would fail on a nil `kubescapeOpsTool`.

It doesn't. `kubescape/helm-release-upgrader`'s `entrypoint.sh` runs a bare `helm upgrade kubescape kubescape/kubescape-operator -n kubescape` with no value flags at all, so it resets to the new chart's defaults — which do include `kubescapeOpsTool`. The first hop is safe.

But it's safe for a different reason than the PR's regression test implies, and that same discovery is what makes the second finding below a real change rather than a refinement.

**Review by amirmalka — APPROVED** [review]
Round 2 — approving. All three findings are addressed, and **one of them was mine to withdraw**.

## I was wrong about the upgrade semantics, and your rebuttal is right

I claimed a bare `helm upgrade` discards prior user overrides, which would have made `ResetThenReuseValues` a behavioural change. You cited the Helm source, so I read it rather than restating my position. `pkg/action/upgrade.go:585-588`:

```go
if len(newVals) == 0 && len(current.Config) > 0 {
	u.cfg.Log("copying values from %s (v%d) to new release.", current.Name, current.Version)
	newVals = current.Config
}
```

With no `-f` and no `--set`, `newVals` is empty, so the default path copies the previous release's config forward — and `reuseValues`' own doc comment at :545-551 states exactly that. Since `up.Run` passes `map[string]any{}`, the default path and `ResetThenReuseValues` resolve to identical values for this caller. There was no behavioural change, the original comment was correct, and the doc section you added is accurate. My apologies for the detour.

One point in the flag's favour that neither of us made: setting it explicitly is real future-proofing rather than a no-op restatement. The default path only preserves overrides *because* `newVals` happens to be empty. If anyone later adds a `--set`-style flag to this subcommand, that condition stops holding and the default path would silently stop carrying `current.Config` forward — `ResetThenReuseValues` keeps doing so unconditionally. Worth a half-sentence in the doc if you touch it again, since it's the durable reason to keep the flag.

## The release guard

Fixed, and placed better than I suggested. `action.NewHistory(cfg).Run()` with `errors.Is(err, driver.ErrReleaseNotFound)` sits **before** `AddRepo`, so a CronJob outliving its release no longer downloads a chart index it has no use for either.

`TestUpgradeSkipsUnknownRelease` is the detail I'd single out: asserting that a missing release is skipped *even when the repo is unreachable* pins the call ordering rather than just the outcome, so a future refactor that moves the guard below `AddRepo` fails the test instead of quietly regressing. `TestUpgradePropagatesHistoryError` covers the inverse with a failing driver, so a genuine history error can't be swallowed as "not found" — that's the pair I'd have asked for.

I also checked the caller, since `Upgrade` can now return `(nil, nil)`: `runHelmUpgrade` does `_, err = helmupgrade.Upgrade(...)` and never touches the release, so there's no nil dereference waiting behind the new contract.

## `--days`

Bounded 1-73000 with the range surfaced in both the flag help and the usage text, validated ahead of the required-flag checks. 73000 days is ~6.3e18 ns against an `int64` ceiling of ~9.22e18, so the overflow path is comfortably unreachable rather than merely unlikely.

## Merge gate

CI is red on **DCO**: `8c06415` is the only commit in the PR missing a `Signed-off-by` line — the other six have one. `git commit --amend -s` and a force-push clears it. Treating that as a merge gate rather than a review item, hence the approval.

Everything from round 1 that I verified still stands: the certgen crypto is stricter than the shell it replaces, the Helm cache dirs are correctly redirected for `readOnlyRootFilesystem`, the chart swap is clean like-for-like, and the publish workflow uses `pull_request` with a same-repo guard rather than `pull_request_target`.
